cpp/nacjd.cc

```diff
@@ -0,0 +1,360 @@
+const unsigned int timeout(15);
+const std::string nacjd_titles = UBTools::GetTuelibPath() + "nacjd_titles.html";
+const std::string nacjd_new_titles_json = UBTools::GetTuelibPath() + "nacjd_new_titles.json";
+
+
+bool ContainsValue(const std::map<std::string,std::string> &map, const std::string search_value) {
+    for (const auto &[key, val] : map) {
+        if (val == search_value)
+            return true;
+    }
+    return false;
+}
+
+
+void HandleChar(const char &c) {
+    if (state == 0) {
+        if (c == '"')
+            state = 1;
+    } else if (state == 1) {
+        if (c == 'I')
+            state = 2;
+        else
+            state = 0;
+    } else if (state == 2) {
+        if (c == 'D')
+            state = 3;
+        else
+            state = 0;
+    } else if (state == 3) {
+        if (c == '"')
+            state = 4;
+        else
+            state = 0;
+    } else if (state == 4) {
+        if (c == ':')
+            state = 5;
+        else
+            state = 0;
+    } else if (state == 5) {
+        if (c == ',') {
+            ids_website.push_back(current_id_website);
+            current_id_website = "";
+            state = 0;
+        }
+        else
+            current_id_website += c;
+    }
+}
+
+
+bool DownloadID(std::ofstream &json_new_titles, const std::string &id, const bool use_separator) {
+    const std::string DOWNLOAD_URL("https://pcms.icpsr.umich.edu/pcms/api/1.0/studies/" + id +"/versions/V1/dats?page=https://www.icpsr.umich.edu/web/NACJD/studies/37644/export&user=");
+
+    Downloader downloader(DOWNLOAD_URL, Downloader::Params(), timeout * 1000);
+    if (downloader.anErrorOccurred()) {
+        LOG_WARNING("Error while downloading data for id " + id + ": " + downloader.getLastErrorMessage());
+        return false;
+    }
+
+     // Check for rate limiting and error status codes:
+    const HttpHeader http_header(downloader.getMessageHeader());
+    if (http_header.getStatusCode() == 429)
+        LOG_ERROR("we got rate limited!");
+    else if (http_header.getStatusCode() != 200) {
+        LOG_WARNING("NACJD returned HTTP status code " + std::to_string(http_header.getStatusCode()) + "! for NACJD id: " + id);
+        return false;
+    }
+
+    const std::string &json_document(downloader.getMessageBody());
+    std::shared_ptr<JSON::JSONNode> full_tree;
+    JSON::Parser parser(json_document);
+    if (not parser.parse(&full_tree)) {
+        LOG_ERROR("failed to parse JSON (" + parser.getErrorMessage() + "), download URL was: " + DOWNLOAD_URL);
+        return false;
+    }
+
+    const std::shared_ptr<const JSON::ObjectNode> top_node(JSON::JSONNode::CastToObjectNodeOrDie("full_tree", full_tree));
+    if (use_separator)
+        json_new_titles << "," << "\n";
+    json_new_titles << top_node->toString() << "\n";
+
+    return true;
+}
+
+
+void ExtractExistingIDsFromMarc(MARC::Reader * const marc_reader, std::set<std::string> &parsed_marc_ids) {
```

Comment:
parsed_marc_ids should be a pointer. [Ask me if you're curious about the rationale.]

Reply:
parsed_marc_ids should be a pointer.  We don't use references, w/ the exception of streams, for mutable parameters.